core/src/main/java/com/linecorp/armeria/client/RetryingClient.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2016 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.util.function.Supplier;
+
+import com.linecorp.armeria.common.Request;
+import com.linecorp.armeria.common.Response;
+
+/**
+ * A {@link Client} decorator that handles failures of remote invocation and retries requests.
+ *
+ * @param <I> the {@link Request} type
+ * @param <O> the {@link Response} type
+ */
+public abstract class RetryingClient<I extends Request, O extends Response>
+        extends DecoratingClient<I, O, I, O> {
+    private final int maxRetries;
+    private final Supplier<? extends Backoff> backoffSupplier;
+
+    /**
+     * Creates a new instance that decorates the specified {@link Client}.
+     */
+    RetryingClient(Client<? super I, ? extends O> delegate,
+                   int maxRetries, Supplier<? extends Backoff> backoffSupplier) {
```

Comment:
Shouldn't `maxRetries` part of `RetryRequestStrategy`? e.g. `RetryRequestStrategy.upTo(3)`